In [2]:
library(ISLR2)
set.seed(1)
library(boot)

### Workgroup 5

- Sonia Rosmery Asto
- Daniel Carrillo
- Nicole Linares

## Bootstraping - Scripts in R 

First, we load the data.

In [3]:
## loading the data
Penn <- as.data.frame(read.table("../data/penn_jae.dat", header=T ))
n <- dim(Penn)[1]
p_1 <- dim(Penn)[2]
Penn<- subset(Penn, tg== 4| tg==0)
attach(Penn)

We select the treatment group number 4

In [4]:
T4<- (tg==4)
summary(T4)

   Mode   FALSE    TRUE 
logical    3354    1745 

In [4]:
head(Penn)

abdt  tg inuidur1 inuidur2 female black hispanic othrace dep q1 ... q5 q6
1  10824 0  18       18       0      0     0        0       2   0  ... 1  0 
4  10824 0   1        1       0      0     0        0       0   0  ... 1  0 
5  10747 0  27       27       0      0     0        0       0   0  ... 0  0 
12 10607 4   9        9       0      0     0        0       0   0  ... 0  0 
13 10831 0  27       27       0      0     0        0       1   0  ... 1  0 
14 10845 0  27       27       1      0     0        0       0   0  ... 1  0 
   recall agelt35 agegt54 durable nondurable lusd husd muld
1  0      0       0       0       0          0    1    0   
4  0      0       0       0       0          1    0    0   
5  0      0       0       0       0          1    0    0   
12 0      1       0       0       0          0    0    1   
13 0      0       1       1       0          1    0    0   
14 0      0       1       0       0          1    0    0

In [5]:
dim(Penn)

[1] 5099   23

We create the `boot.fn` function to create bootstrap estimates. We will use this function not only to estimate the bootstrap estimation, but also to estimate the function with the linear regression and compare the values of the coefficients and standard errors. 

As the number of observations is 5099, the samples will be according to this number. The function will allow us to compute the estimates of the betas associated to the variables. 

In [6]:
boot.fn<- function(data, index)
    coef(lm(log(inuidur1)~ T4 + (female+black+othrace+factor(dep)+q2+q3+q4+q5+q6+agelt35+agegt54+durable+lusd+husd), data=data, subset=index))

In [7]:
lm(log(inuidur1)~ T4 + (female+black+othrace+factor(dep)+q2+q3+q4+q5+q6+agelt35+agegt54+durable+lusd+husd),data=Penn)


Call:
lm(formula = log(inuidur1) ~ T4 + (female + black + othrace + 
    factor(dep) + q2 + q3 + q4 + q5 + q6 + agelt35 + agegt54 + 
    durable + lusd + husd), data = Penn)

Coefficients:
 (Intercept)        T4TRUE        female         black       othrace  
    2.178462     -0.071692      0.126368     -0.293768     -0.472445  
factor(dep)1  factor(dep)2            q2            q3            q4  
    0.029867      0.096187      0.073678     -0.038507     -0.054949  
          q5            q6       agelt35       agegt54       durable  
   -0.144178      0.003361     -0.162772      0.229667      0.126557  
        lusd          husd  
   -0.175353     -0.105225  


In [8]:
boot.fn(Penn, 1:5099)

(Intercept)       T4TRUE       female        black      othrace factor(dep)1 
 2.178462326 -0.071692484  0.126368328 -0.293767980 -0.472445058  0.029866899 
factor(dep)2           q2           q3           q4           q5           q6 
 0.096186517  0.073678072 -0.038506537 -0.054949195 -0.144177912  0.003361318 
     agelt35      agegt54      durable         lusd         husd 
-0.162772168  0.229666708  0.126557359 -0.175352572 -0.105224727

In [9]:
sample(5099,5099,replace=T)

[1] 1017 4775 2177 5026 1533 4567 2347  270 4050 3379 4065  597 1301  330
  [15] 1799 3913 1749   37 1129  729  878  485 4012 5071 2849 2900 2378 4650
  [29] 1446 2159 3476 1948 2580 1530 4439 4136 4633 4344 1222 2426 2087 2483
  [43] 2858 1696  526 1069   22 1128  983 1791 3910 1639 4939  465 5092 1200
  [57] 3863 1134   84 1895 3101 2300 3990 1948 1328  557  287 3217 3702 1522
  [71]  858 4672  990 3175  316  733 4907 5051 2330 1167 3514 3992 1706  501
  [85] 3788  536 5084 3286 3747   29 3662 1942 4934 1820 2281 1317 4669 1966
  [99]  369 2499 4182  355 1073  361 1340 1266 1841 2866 4343  751  219  135
 [113] 4207  532 4504 3123  912 3428 2178 4455 2153 1148 1101 1242 3682 1218
 [127] 4115  273  418  867 4782 4499 3821 1611  818 2652 4730  664 3210  719
 [141]  500 3045 2809  423  421  989 4236 4222 3598 3580 5094 3700 2319 1154
 [155] 2625 3523  504 4454  785 3796 3912 1572 4401 1833 2461 5077 2624 4225
 [169]  309 2922 4078  441 2904 3189 4405  470 3686 1360 1822 1790  349 3144
 [183]  894 4686 1956  474  421 5048  455 3700 3306 3556 3193   15 3366 4158
 [197] 4486 1668 1059 4477 1129 2597 4096 2012 4084 2110 1172  797 2858 1596
 [211] 4072 5057 4430 2079 3372 2762  610 3354 1265 3068 4182 2265 2840 2156
 [225]  751 1760 4664 5098 4839 4059 4075 3506 2964 3421 3586 4212 2043 3070
 [239] 1463 1221 3292  462 1957 2283 4049 1498 2221 3937 1498 4503 2033  731
 [253] 4281 2813 3805 4589 4771 3914  966 2492 1191 2017 3774  291 3979 4749
 [267]  247 2482 3388 3418 2984 2722 2884   56  934 3153 4425 4590 4425 2966
 [281] 3552 2152 4257 2432  127 4356  717 4156 2496 4277 1452 3997 4387  676
 [295] 1993 4707  148 4246  121  169  598 1491 4204 2815 2578 1580 3869 1115
 [309] 3616 3551 4147 2167 2513 1783  436 2137  533 4092 1381 2752  846  219
 [323]  858 3486 3565  570 1926 1813 2833 3189 1498 1858 4729  281 4063 4730
 [337]  724 2053 1354  492 3255 4664 1949  514 2117 4666 1253 2016 2189 2841
 [351]  132 5094 1659 2802 2181 4846 1421 2312 2643 5040 1839 3505 1838 1766
 [365] 1544  320 4948 4211 3525 1474 1242 4978 3917 2167 2543 2018 1822  109
 [379] 3760 2587  753 4907  393  648 4141 4208 5036 4318 4254 4225 2822 1937
 [393] 3388 3789 5046 2778 2016 3940 4396 3333 1139 4436 4508   99 3200 3822
 [407] 2542 4879 3042 1190 1459 1776  116 2948 3157 1749 1326 1847 1232 1018
 [421]  383  728 3349  771 2906  484 4599 3222 3887 2563 2800 3588 1700   58
 [435] 3380 1509 4801 3183 2021 4438 2170 4140 3700 1803 1414 3120 1124  313
 [449] 2004  822 3787 3382 3157   81 3328 3126 2194 2796 2775 1401  435 3330
 [463] 1786 3360  234 4392 4411 3926 4813 1779 2178 3471 2615 4278   61 4606
 [477] 1089  649 3505 1861  219 2114 2649 2775 2992 3452 3797 4549 1177   17
 [491] 1057 3706 2257 1525 3415 4720 2079  795  573 3037 3298 2744 3558 4568
 [505] 2351 3651 4898 2298 1871 1694 5034 5005  363 3236  624 3932 2573 1267
 [519] 4690 3329 2373  702 1378  959 4378 2273 4115 1489  397 5031 3044 4182
 [533]  744  564 4574 1592 2290 3482 3459   87 4313  860 2446  628 2399 1104
 [547] 1493 4140 4062 3926 1518 4784 2811 1473 2572 4572 4364 1177 4402 2137
 [561]  671 4933 2569 1218 4745 2551  128 3497 3350 1435  102 3644 4066 3004
 [575] 3457 2370 2678 2973  453 2402 4585 3468 1129  815  574 4109  644 4402
 [589]   48 3437 1174 2608 3854 1529 4358 1590 4693 3352  835 2280 1550 4123
 [603] 2899 1821  573 4165  601 1759  718 2107 4343 1019 2020 3924  543 2563
 [617]  296 1930 4557  764 1734 1354  122 3047  498 1043  376 4562 3255 2734
 [631]  919 1242 1487 1366 4673  854  525 1814 2490  108 4137 2145 3862 4221
 [645] 3018 2285 4435 2056  396 2507  144 1835 3733 3290  252 2504 2069 3591
 [659] 4963 2346 4897  133 1803 1641  386 1802  290  770  902 2158 4666 4297
 [673]  487 2053 2513 3254 1421  908 3363 1800 3714 4286  247 2362  457  991
 [687] 3130 4222  583 4279  647 4007 2732  492 3807 3586 1620 3129  553 1780
 [701] 1864 4633 1443 1787 2037 4030 3408 1031 2265 4330 3027 2298 2531 3417
 [715] 2573 2608 1587 4590 1651 3402 1630  672 2136 4364 4488  151 1524 4085

In [10]:
set.seed(1)
boot.fn(Penn, sample(5099,5099,replace=T))

(Intercept)       T4TRUE       female        black      othrace factor(dep)1 
 2.258670234 -0.117228952  0.142462443 -0.374373286 -0.713971846  0.009594995 
factor(dep)2           q2           q3           q4           q5           q6 
 0.090434910  0.005622350 -0.133365893 -0.114376148 -0.180682473  0.036319276 
     agelt35      agegt54      durable         lusd         husd 
-0.153838318  0.185651787  0.085436134 -0.156716042 -0.092207747

With the function `boot()` we will compute the standard errors of 1,000 bootstrap estimates for $\beta_0$ (intercept) and for all the other variables. 

In [11]:
alo <- boot(Penn, boot.fn, 1000)

In [12]:
alo


ORDINARY NONPARAMETRIC BOOTSTRAP


Call:
boot(data = Penn, statistic = boot.fn, R = 1000)


Bootstrap Statistics :
         original        bias    std. error
t1*   2.178462326  0.0010812733  0.15469174
t2*  -0.071692484 -0.0003483223  0.03552280
t3*   0.126368328  0.0007853971  0.03518438
t4*  -0.293767980  0.0007320124  0.06028143
t5*  -0.472445058 -0.0049861202  0.24369741
t6*   0.029866899  0.0007690384  0.05555582
t7*   0.096186517 -0.0011770733  0.04531520
t8*   0.073678072 -0.0010033034  0.15235277
t9*  -0.038506537 -0.0003146298  0.15005091
t10* -0.054949195 -0.0008440703  0.15097113
t11* -0.144177912 -0.0015435867  0.15021731
t12*  0.003361318 -0.0038177923  0.16140880
t13* -0.162772168  0.0017786902  0.03816477
t14*  0.229666708  0.0002651347  0.05836239
t15*  0.126557359  0.0001236963  0.04853387
t16* -0.175352572 -0.0006751971  0.04095015
t17* -0.105224727 -0.0007444597  0.04512871

In [13]:
ols <- lm(log(inuidur1)~ T4 + (female+black+othrace+factor(dep)+q2+q3+q4+q5+q6+agelt35+agegt54+durable+lusd+husd))

In [14]:
ols_1 <-summary(ols, data = Penn)$coef
ols_1

Estimate     Std. Error t value     Pr(>|t|)    
(Intercept)   2.178462326 0.15901507 13.69972271 5.631260e-42
T4TRUE       -0.071692484 0.03546326 -2.02159887 4.327012e-02
female        0.126368328 0.03482493  3.62867401 2.876799e-04
black        -0.293767980 0.05297556 -5.54534899 3.081966e-08
othrace      -0.472445058 0.19839750 -2.38130547 1.728801e-02
factor(dep)1  0.029866899 0.05414025  0.55165799 5.812070e-01
factor(dep)2  0.096186517 0.04686228  2.05253621 4.016868e-02
q2            0.073678072 0.15682593  0.46980799 6.385124e-01
q3           -0.038506537 0.15647802 -0.24608272 8.056281e-01
q4           -0.054949195 0.15656019 -0.35097809 7.256193e-01
q5           -0.144177912 0.15588797 -0.92488160 3.550714e-01
q6            0.003361318 0.16645573  0.02019347 9.838898e-01
agelt35      -0.162772168 0.03696505 -4.40340780 1.087509e-05
agegt54       0.229666708 0.05917173  3.88135887 1.051906e-04
durable       0.126557359 0.04814585  2.62862426 8.598901e-03
lusd         -0.175352572 0.04098250 -4.27871860 1.914271e-05
husd         -0.105224727 0.04490864 -2.34308443 1.916341e-02

The results between the bootstrap approach and the linear regression are slightly different. The standard error from the bootstrap estimation are a little bit higher with the variables $T4$, $female$ and $black$. As we can see, in the linear estimation the standard errors for those variables are $0.03546326$, $0.03482493$ and $0.05297556$. And for the other method are $0.03420464$, $0.03540692$ and $0.057385124$. We can trust more in the bootstraping method, even tho the std's are higher, since it doesn't depends on assumptions that linear regression does. 

In [15]:
table<- matrix(0, 2, 4)

table[1,1]<-ols_1[1,2]
table[1,2]<-ols_1[2,2]
table[1,3]<-ols_1[3,2]
table[1,4]<-ols_1[4,2]

table[2,1]<- 0.15511158
table[2,2]<- 0.03420464
table[2,3]<- 0.03540692
table[2,4]<- 0.05738512

colnames(table) <- c('intercept','T4TRUE','female','black')
rownames(table)<- c("linear_std", 'boot_std')

table

intercept T4TRUE     female     black     
linear_std 0.1590151 0.03546326 0.03482493 0.05297556
boot_std   0.1551116 0.03420464 0.03540692 0.05738512

## Part 2
The second part of the R script (Causal Tree - Only script in R), can be found in the file Group5_lab5_R_2. The work was divided in two parts because there were problems with the code when both parts were put together.